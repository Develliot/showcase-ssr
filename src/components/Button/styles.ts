import styled from 'styled-components';

export const ButtonStyled = styled.button`
    font-size: ${({
        theme: {
            fontSizes: { medium },
        },
    }) => medium};
    padding: ${({
        theme: {
            spacing: { tiny, medium },
        },
    }) => `${tiny} ${medium}`};
    font-weight: 500;
    text-transform: uppercase;
    transition: color 0.3s ease, background-color 0.3s ease;
    background-color: ${({
        color,
        theme: {
            colors,
            colors: { blue },
        },
    }) => (color ? colors[`${color}`] : blue)};
    border-width: 0.12rem;
    border-style: solid;
    border-radius: 1.5rem;
    border-color: ${({
        color,
        theme: {
            colors,
            colors: { blue },
        },
    }) => (color ? colors[`${color}`] : blue)};
    color: ${({
        color,
        theme: {
            colors: { white, darkGrey },
        },
    }) => {
        switch (color) {
            case 'white':
                return darkGrey;
            case 'yellow':
                return darkGrey;
            case 'blue':
                return white;
            case 'pink':
                return white;
            case 'darkGrey':
                return white;
            case undefined:
                return white;
            default:
                return white;
        }
    }};

    &:hover,
    &:focus {
        background-color: ${({
            color,
            theme: {
                colors: { white },
            },
        }) => {
            switch (color) {
                case 'white':
                    return `rgba(0, 0, 0, 0.4)`;
                case 'yellow':
                    return `rgba(0, 0, 0, 0.4)`;
                case 'blue':
                    return white;
                case 'pink':
                    return white;
                case 'darkGrey':
                    return white;
                case undefined:
                    return white;
                default:
                    return white;
            }
        }};
        color: ${({
            color,
            theme: {
                colors,
                colors: { blue },
            },
        }) => (color ? colors[`${color}`] : blue)};
    }
`;
