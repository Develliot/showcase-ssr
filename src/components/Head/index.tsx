import Head from 'next/head';
import { colors } from 'src/theme';
import { useRouter } from 'next/router';

const nodeEnv = process.env['NODE_ENV'];

const baseUrls = {
    production: 'https://showcase-ssr.appspot.com',
    development: 'http://localhost:3000',
    test: 'http://localhost:3000',
};

const baseUrl = baseUrls[nodeEnv];

type Props = {
    title: string;
    description: string;
    imageUrl?: string;
    imageAlt?: string;
};

const HeadComponent = ({
    title,
    description,

    imageUrl,
    imageAlt,
}: Props) => {
    const url = baseUrl + useRouter().asPath;
    return (
        <Head>
            {/* General */}
            <title>{title}</title>
            <meta name='description' content={description} />
            <base href={baseUrl} />
            <meta name='theme-color' content={colors.blue} />

            {/* Facebook */}
            <meta property='og:title' content={title} />
            <meta property='og:type' content='website' />
            <meta property='og:url' content={url} />
            <meta property='og:description' content={description} />
            <meta property='og:site_name' content='SSR Showcase' />
            <meta property='og:locale' content='en_EN' />
            <meta property='article:author' content='Elliot Revan' />
            {!!imageUrl && <meta property='og:image' content={imageUrl} />}
            {!!imageAlt && <meta property='og:image:alt' content={imageAlt} />}

            {/* Twitter */}
            <meta name='twitter:card' content='summary' />
            <meta name='twitter:url' content={url} />
            <meta name='twitter:title' content={title} />
            <meta name='twitter:description' content={description} />
            {!!imageUrl && <meta property='twitter:image' content={imageUrl} />}
            {!!imageAlt && <meta property='og:image:alt' content={imageAlt} />}

            {/* <meta name='twitter:site' content='@site_account' /> */}
            {/* <meta name='twitter:creator' content='@individual_account' /> */}
        </Head>
    );
};

export default HeadComponent;
