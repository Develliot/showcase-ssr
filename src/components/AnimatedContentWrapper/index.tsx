import React, { useState, useRef, FunctionComponent } from 'react';
import { useScrollPosition } from 'src/hooks/useScrollPosition';
import { ContentWrapperStyled } from './styles';

export const AnimatedContentWrapper: FunctionComponent = ({ children }) => {
    const [hideOnScroll, setHideOnScroll] = useState(true);
    const wrapperRef = useRef<HTMLDivElement>(null);
    const isBrowser = typeof window !== `undefined`;

    useScrollPosition(
        ({
            currPos,
        }: {
            prevPos: { x: number; y: number };
            currPos: { x: number; y: number };
        }) => {
            const isShow = isBrowser ? currPos.y < window.innerHeight : false;
            if (isShow !== hideOnScroll) setHideOnScroll(isShow);
        },
        [hideOnScroll],
        wrapperRef,
        false,
        300
    );

    return (
        <ContentWrapperStyled ref={wrapperRef} visible={hideOnScroll}>
            {children}
        </ContentWrapperStyled>
    );
};

export default AnimatedContentWrapper;
