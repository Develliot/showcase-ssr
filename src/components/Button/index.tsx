import React, { FunctionComponent } from 'react';
import { ColorType } from 'src/theme';

import { ButtonStyled } from './styles';

type Props = {
    color?: ColorType;
    onClick?: () => void;
};

export const Button: FunctionComponent<Props> = ({
    color = 'blue',
    onClick = () => null,
    children,
}) => {
    return (
        <ButtonStyled onClick={onClick} color={color}>
            {children}
        </ButtonStyled>
    );
};

export default Button;
