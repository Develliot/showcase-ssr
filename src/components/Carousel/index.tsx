import React, { FunctionComponent, useState, useRef } from 'react';
import _ from 'underscore';

import {
    CarouselWrapper,
    CarouselCardInner,
    CarouselCard,
    CarouselScrollArea,
    CarouselFadeOutLeft,
    CarouselFadeOutRight,
    LeftButtonContainer,
    RightButtonContainer,
    CarouselButton,
} from './styles';
import { VerticalSpacer } from 'src/components/Layout';
import { Dots } from './Dots';

type Props = {
    onPositionChanged?: (position: number) => void;
};

export const CarouselLeftButton = ({ onClick }: { onClick: () => void }) => (
    <LeftButtonContainer>
        <CarouselButton
            onClick={onClick}
            aria-label='previous user'
        >{`◀`}</CarouselButton>
    </LeftButtonContainer>
);

export const CarouselRightButton = ({ onClick }: { onClick: () => void }) => (
    <RightButtonContainer>
        <CarouselButton
            onClick={onClick}
            aria-label='next user'
        >{`▶`}</CarouselButton>
    </RightButtonContainer>
);

export const Carousel: FunctionComponent<Props> = ({
    children,
    onPositionChanged,
}) => {
    const [currentPosition, setCurrentPosition] = useState(0);
    const scrollRef = useRef<HTMLDivElement>(null);

    // remove null children
    const childArray = React.Children.toArray(children).filter(Boolean);

    const getScroll = () => {
        return {
            y: scrollRef.current ? scrollRef.current.scrollTop : undefined,
            x: scrollRef.current ? scrollRef.current.scrollLeft : undefined,
        };
    };

    const setPosition = (to: number) => {
        setCurrentPosition(to);

        if (onPositionChanged) {
            onPositionChanged(to);
        }
    };

    const scrollToPosition = (to: number) => {
        if (scrollRef && scrollRef.current) {
            const rect = scrollRef.current.getBoundingClientRect();
            const width = rect.width;
            const scrollPosition = Math.round(width * 0.8 * to);
            scrollRef.current.scrollTo(scrollPosition, 0);
        }
        setPosition(to);
    };

    const handleScroll = () => {
        if (!scrollRef || !scrollRef.current) {
            return null;
        }
        const rect = scrollRef.current.getBoundingClientRect();
        const width = rect.width;
        const scrollPosition = getScroll().x;
        if (scrollPosition) {
            const currentPosition = Math.round(scrollPosition / (width * 0.8));
            setPosition(currentPosition);
        }
    };

    const handleScrollThrottled = _.throttle(handleScroll, 100);

    const moreOnRight = currentPosition !== childArray.length - 1;
    const moreOnLeft = currentPosition !== 0;

    return (
        <>
            <CarouselWrapper>
                <CarouselScrollArea
                    ref={scrollRef}
                    onScroll={handleScrollThrottled}
                >
                    {childArray.map((child, index) => {
                        const isSelected = index === currentPosition;
                        return (
                            <CarouselCard key={index}>
                                <CarouselCardInner isSelected={isSelected}>
                                    {child}
                                </CarouselCardInner>
                            </CarouselCard>
                        );
                    })}
                </CarouselScrollArea>
                <CarouselFadeOutLeft visible={moreOnLeft} />
                <CarouselFadeOutRight visible={moreOnRight} />
                {moreOnLeft ? (
                    <CarouselLeftButton
                        onClick={() => {
                            scrollToPosition(Math.max(0, currentPosition - 1));
                        }}
                    />
                ) : null}
                {moreOnRight ? (
                    <CarouselRightButton
                        onClick={() => {
                            scrollToPosition(
                                Math.min(
                                    currentPosition + 1,
                                    childArray.length - 1
                                )
                            );
                        }}
                    />
                ) : null}
            </CarouselWrapper>
            <VerticalSpacer size='small' />
            <Dots
                count={childArray.length}
                currentPosition={currentPosition}
                scrollToPosition={scrollToPosition}
            />
        </>
    );
};

export default Carousel;
