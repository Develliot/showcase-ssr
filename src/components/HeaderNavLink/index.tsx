import React from 'react';
import { Paragraph } from 'src/components/Typography';
import { RowCenter } from 'src/components/Layout';
import { NavLinkStyled, HoverInidcator, HeaderLinkItem } from './styles';
import Link from 'next/link';

export const HeaderNavLink = ({
    to,
    title,
    selected,
}: {
    to: string;
    title: string;
    selected: boolean;
}) => {
    return (
        <HeaderLinkItem>
            <Link href={to} passHref>
                <NavLinkStyled title={title} tabIndex={0}>
                    <Paragraph color='white'>{title}</Paragraph>
                    <RowCenter>
                        <HoverInidcator selected={selected} />
                    </RowCenter>
                </NavLinkStyled>
            </Link>
        </HeaderLinkItem>
    );
};
