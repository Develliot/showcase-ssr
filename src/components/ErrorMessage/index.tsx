import React, { FunctionComponent } from 'react';
import Link from 'next/link';

import urls from 'src/urls';

import { Button } from 'src/components/Button';
import { H2, Paragraph } from 'src/components/Typography';
import { VerticalSpacer } from 'src/components/Layout';

import { ErrorWrapper } from './styles';

type Props = {
    errorMessage: string;
    retry?: () => void;
};

export const ErrorMessage: FunctionComponent<Props> = ({
    errorMessage,
    retry,
}) => {
    return (
        <ErrorWrapper>
            <H2>Something broke!</H2>
            <VerticalSpacer size='large' />
            <Paragraph>{errorMessage}</Paragraph>
            <VerticalSpacer size='large' />
            {retry ? (
                <>
                    <Button onClick={retry}>Retry</Button>
                    <VerticalSpacer size='medium' />
                </>
            ) : null}
            <Link href={urls.home}>
                <a>
                    <Button color='darkGrey'>Home</Button>
                </a>
            </Link>
        </ErrorWrapper>
    );
};

export default ErrorMessage;
