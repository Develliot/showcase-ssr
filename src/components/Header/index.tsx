import React, { useState } from 'react';
import { useRouter } from 'next/router';

import { useScrollPosition } from 'src/hooks/useScrollPosition';
import urls from 'src/urls';

import { H2 } from 'src/components/Typography';
import { VerticalSpacer } from 'src/components/Layout';
import { HeaderWrapper, HeaderInner, Links } from './styles';

import { HeaderNavLink } from 'src/components/HeaderNavLink';

export const Header = ({
    visible,
    currentRoute,
}: {
    visible: boolean;
    currentRoute: string;
}) => {
    const isSelected = (url: string): boolean => url === currentRoute;

    return (
        <HeaderWrapper visible={visible}>
            <HeaderInner>
                <H2 color='white'>Showcase</H2>
                <VerticalSpacer size='medium' />
                <nav>
                    <Links>
                        <HeaderNavLink
                            to={urls.home}
                            title='Home'
                            selected={isSelected(urls.home)}
                        />
                        <HeaderNavLink
                            to={urls.designSystem}
                            title='Design System'
                            selected={isSelected(urls.designSystem)}
                        />
                        <HeaderNavLink
                            to={urls.usingAPIs}
                            title='Using APIs'
                            selected={isSelected(urls.usingAPIs)}
                        />
                    </Links>
                </nav>
            </HeaderInner>
        </HeaderWrapper>
    );
};

export const HideOnScroll = () => {
    const [hideOnScroll, setHideOnScroll] = useState(true);
    const router = useRouter();

    useScrollPosition(
        ({
            prevPos,
            currPos,
        }: {
            prevPos: { x: number; y: number };
            currPos: { x: number; y: number };
        }) => {
            const isShow = currPos.y >= prevPos.y;
            if (isShow !== hideOnScroll) setHideOnScroll(isShow);
        },
        [hideOnScroll],
        null,
        false,
        300
    );

    return <Header visible={hideOnScroll} currentRoute={router.route} />;
};

export default HideOnScroll;
