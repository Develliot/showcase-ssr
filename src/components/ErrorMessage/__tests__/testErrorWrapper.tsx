import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { ErrorMessage } from '../';
import { Button } from 'src/components/Button';

it('renders the with an error message', () => {
    const wrapper: ShallowWrapper = shallow(
        <ErrorMessage errorMessage='This is an error message' />
    );
    expect(wrapper).toMatchSnapshot();
});

it('renders the with a retry button', () => {
    const mockRetry = jest.fn();
    const wrapper: ShallowWrapper = shallow(
        <ErrorMessage
            errorMessage='This is an error message'
            retry={mockRetry}
        />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Button)).toHaveLength(2);
    wrapper
        .findWhere(
            element => element.type() === Button && element.contains('Retry')
        )
        .simulate('click');
    expect(mockRetry).toHaveBeenCalledTimes(1);
});
