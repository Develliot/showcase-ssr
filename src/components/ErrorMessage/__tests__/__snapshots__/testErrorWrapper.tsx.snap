// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the with a retry button 1`] = `
<styles__ErrorWrapper>
  <Typography__H2>
    Something broke!
  </Typography__H2>
  <Layout__VerticalSpacer
    size="large"
  />
  <Typography__Paragraph>
    This is an error message
  </Typography__Paragraph>
  <Layout__VerticalSpacer
    size="large"
  />
  <Button
    onClick={[MockFunction]}
  >
    Retry
  </Button>
  <Layout__VerticalSpacer
    size="medium"
  />
  <Link
    href="/"
    passHref={true}
  >
    <a>
      <Button
        color="darkGrey"
      >
        Home
      </Button>
    </a>
  </Link>
</styles__ErrorWrapper>
`;

exports[`renders the with an error message 1`] = `
<styles__ErrorWrapper>
  <Typography__H2>
    Something broke!
  </Typography__H2>
  <Layout__VerticalSpacer
    size="large"
  />
  <Typography__Paragraph>
    This is an error message
  </Typography__Paragraph>
  <Layout__VerticalSpacer
    size="large"
  />
  <Link
    href="/"
    passHref={true}
  >
    <a>
      <Button
        color="darkGrey"
      >
        Home
      </Button>
    </a>
  </Link>
</styles__ErrorWrapper>
`;
